### PR TITLE
`split_data` function

### PR DESCRIPTION
**Pull request by jwa7**
This function wraps the `split` and `unique_metadata` operations in equistore, allowing users to split multiple TensorMaps according to named indices (randomly shuffled or otherwise) along a given TensorBlock axis. It is high-level enough such that it does not require the specification of any Labels objects.

**Example Usage**

```py
import equistore.io
from equisolve.utils import split_data

# Load input and output TensorMaps
input = equistore.io("/path/to/input/tensormap/")
output = equistore.io("/path/to/output/tensormap/")

[[in_train, in_test], [out_train, out_test]], grouped_idxs = split_data(
            tensors=[input, output],
            axis="samples", 
            names=["structure"],
            n_groups=2,  # for train-test split
            group_sizes_rel=[0.8, 0.2],  # 80:20 train:test split
            random_seed=100,
        )
```

The function returns the split tensors, where each respective tensor in each group has the same unique metadata indices. In other words, in the example above, `in_train` and `out_train` will have the same unique "structure" metadata indices along the "samples" axes of each block. `in_test` and `out_test` will also have the same indices as eachother, but be distinct (zero intersection) with the indices in the train TensorMaps.

The function also returns a list of Labels object for the grouped indices (`grouped_idxs` in the ex. above). These show the `unique_metadata` for the specified `axis` and `names` present in each of the groups of split tensors.

----
:books: Documentation preview :books:: https://equisolve--25.org.readthedocs.build/en/25/

### CODE REVIEW

File: src/equisolve/utils/split_data.py [hunk 1]
```diff
@@ -0,0 +1,431 @@
...
+
+        from equisolve.utils import split_data
+
+        [[in_train, in_test], [out_train, out_test]], grouped_labels = split_data(
+            tensors=[input, output],
+            axis="samples",
+            names=["structure"],
+            n_groups=2,                  # for train-test split
+            group_sizes=[0.8, 0.2],  # relative, a 80% 20% train-test split
+            random_seed=100,
```
Comment (PicoCentauri) [inline src/equisolve/utils/split_data.py:132]: Isn't this an overkill if we also can give a list of tensors. It might be simpler for our code and it is not a big burden if the user just has to call the method twice.

File: src/equisolve/utils/split_data.py [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+    _shuffle_indices(unique_idxs, random_seed)
+
+    # Must be at least as many unique indices as groups
+    n_indices = len(unique_idxs)
+    if n_indices < n_groups:
+        raise ValueError(
+            f"the number of groups specified ({n_groups}) is greater than the"
+            + f" number of unique metadata indices ({n_indices}) for the"
+            + f" chosen axis {axis} and names {names}: {unique_idxs}"
+        )
```
Comment (PicoCentauri) [inline src/equisolve/utils/split_data.py:213]: You do not have to stack them with a `+`. Python can do the magic for us here.

```suggestion
        raise ValueError(
            f"the number of groups specified ({n_groups}) is greater than the"
            f" number of unique metadata indices ({n_indices}) for the"
            f" chosen axis {axis} and names {names}: {unique_idxs}"
        )
```

File: src/equisolve/utils/split_data.py [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+    return split_tensors, grouped_labels
+
+
+def _shuffle_indices(indices: Labels, random_seed: Optional[int] = None):
+    """
+    Shuffles the input Labels object `indices` according to the random seed. If
+    `random_seed=None`, not shuffling is performed and the indices are just
+    returned. If -1, shuffling occurs but with no random seed set. If an integer
+    != -1, shuffling is executed but with a random seed set to this value. There
+    is no return value, as the input `indices` is shuffled in place.
```
Comment (PicoCentauri) [inline src/equisolve/utils/split_data.py:244]:
```suggestion
    Inplace shuffling of a `Labels` instance.
    
    If `random_seed=None`, no shuffling is performed and the indices are just
    returned. If `-1`, shuffling with a random seed is performed. If an integer
    != -1, shuffling is executed with the given random seed. 
```

File: src/equisolve/utils/split_data.py [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+
+    # Split each of the input TensorMaps
+    split_tensors = []
+    for tensor in tensors:
+        split_tensors.append(equistore.split(tensor, axis, grouped_labels))
+
+    return split_tensors, grouped_labels
+
+
+def _shuffle_indices(indices: Labels, random_seed: Optional[int] = None):
```
Comment (PicoCentauri) [inline src/equisolve/utils/split_data.py:238]: I would rather say `None` means the function is picking a random number and `-1` does not make sense. Why should one call a shuffle method if one doesn't want to shuffle the indices?

Also I would just call the parameter `seed`.

File: src/equisolve/utils/split_data.py [hunk 5]
```diff
@@ -0,0 +1,431 @@
...
+    """
+    Shuffles the input Labels object `indices` according to the random seed. If
+    `random_seed=None`, not shuffling is performed and the indices are just
+    returned. If -1, shuffling occurs but with no random seed set. If an integer
+    != -1, shuffling is executed but with a random seed set to this value. There
+    is no return value, as the input `indices` is shuffled in place.
+    """
+    if random_seed is not None:  # shuffle
+        if random_seed != -1:  # set a numpy random seed
+            np.random.seed(random_seed)
```
Comment (PicoCentauri) [inline src/equisolve/utils/split_data.py:248]: This is deprecated from numpy instead one should do something like

```python
rng = np.random.default_rng(seed)
rng.shuffle(indices)
```

File: src/equisolve/utils/split_data.py [hunk 6]
```diff
@@ -0,0 +1,431 @@
...
+    _check_args(tensors, axis, names, n_groups, group_sizes, random_seed)
+
+    # Get array of unique indices to split by for each tensor in `tensors`
+    unique_idxs_list = [
+        equistore.unique_metadata(tensor, axis, names) for tensor in tensors
+    ]
+
+    # Check that the unique indices are equivalent for all input tensors
+    _check_labels_equivalent(unique_idxs_list)
+    unique_idxs = unique_idxs_list[0]
```
Comment (PicoCentauri) [inline src/equisolve/utils/split_data.py:201]: Why are you picking the first one here?

File: src/equisolve/utils/split_data.py [hunk 7]
```diff
@@ -0,0 +1,431 @@
...
+
+
+def _get_group_sizes(
+    n_groups: int,
+    n_indices: int,
+    group_sizes: Optional[Union[List[float], List[int]]] = None,
+) -> np.ndarray:
+    """
+    Parses the `group_sizes` arg from :py:func:`split_data` and returns an array
+    of group sizes in absolute numbers.
```
Comment (PicoCentauri) [inline src/equisolve/utils/split_data.py:259]: I don't understand from the docstring what the function is doing. Sorry 😰.

File: src/equisolve/utils/split_data.py [hunk 8]
```diff
@@ -0,0 +1,431 @@
...
+
+    # Check that the sum is preserved
+    assert round(np.sum(array)) == round(np.sum(rounded_array))
+
+    return np.array(rounded_array)
+
+
+def _group_indices(indices: Labels, group_sizes: List[int]) -> Labels:
+    """
+    Sets a numpy `random_seed` then randomly shuffles `indices`. Next, these
```
Comment (PicoCentauri) [inline src/equisolve/utils/split_data.py:317]: There is nothing shuffled in this function or?

File: tests/equisolve_tests/utils/test_split_data.py [hunk 9]
```diff
@@ -0,0 +1,834 @@
...
+
+from equisolve.utils import split_data
+
+
+@pytest.fixture
+def test_tensor_map_a():
+    """
+    Create a dummy tensor map to be used in tests. This is the same one as the
+    tensor map used in `tensor.rs` tests.
+    """
```
Comment (PicoCentauri) [inline tests/equisolve_tests/utils/test_split_data.py:24]: We have to generate some test tensormaps. It is super hard to read the actual tests when they are blown up with these lines.

File: tests/equisolve_tests/utils/test_split_data.py [hunk 10]
```diff
@@ -0,0 +1,834 @@
...
+            assert equistore.equal(i, j)
+
+
+class TestSplitDataErrors:
+    def test_split_data_errors_arg_tensors(self, test_tensor_map_a):
+        """
+        Checks for exceptions on the `tensors` arg.
+        """
+        # Not passing a TensorMap
+        with pytest.raises(TypeError) as e:
```
Comment (PicoCentauri) [inline tests/equisolve_tests/utils/test_split_data.py:584]: You may use the `match` parameter here. This saves a couple of line because you do not have to do an extra assert as you done below.

**Review by PicoCentauri — CHANGES_REQUESTED** [review]
very very helpful function. I look forward using it. I left some comments. Also, good tests!

Most importantly I think it is enough if `split_data` takes a single `TensorMap` or what is your use case that it takes a list?

**Comment by jwa7** [comment]
Hey @PicoCentauri, thanks for all the feedback! I will go through the comments and make the appropriate changes. There is one place in which I disagree, however. The reason that the function takes a list of TensorMaps is twofold:

1) It ensures that each respective TensorMap is split into new TensorMaps where the specified metadata (i.e. "structure" index in the "samples") is the same for each of the split TensorMaps. This is important because if you want to do a train-test split on X and Y, you want X_train and Y_train to have the same structure indices.

2) This more closely matches the functionality of sklearn's train_test_split

Converting the argument to ONLY taking a single TensorMap to me defeats the point of the function - the user wants to perform a split, perhaps with shuffled indices, but doesn't want to have to worry about touching the Labels objects and telling any function which Labels should go where (like they would have to for `split`). This function makes it easy - split my X and Y for me, shuffle the structure indices, and give me back X/Y train/test with all indices in harmony!

File: src/equisolve/utils/split_data.py (hunk 1 above)
Comment (jwa7) [reply src/equisolve/utils/split_data.py:132]: (Commented on below)

File: src/equisolve/utils/split_data.py (hunk 2 above)
Comment (jwa7) [reply src/equisolve/utils/split_data.py:213]: Ah lovely thanks!

File: src/equisolve/utils/split_data.py (hunk 6 above)
Comment (jwa7) [reply src/equisolve/utils/split_data.py:201]: Because line 200 checks that all the Labels objects in the list are equivalent (as they should be), so we only need the first element in this list to use to split by later

File: src/equisolve/utils/split_data.py (hunk 7 above)
Comment (jwa7) [reply src/equisolve/utils/split_data.py:259]: Added more explanation in the docstring

File: src/equisolve/utils/split_data.py (hunk 8 above)
Comment (jwa7) [reply src/equisolve/utils/split_data.py:317]: Good spot!

File: tests/equisolve_tests/utils/test_split_data.py (hunk 9 above)
Comment (jwa7) [reply tests/equisolve_tests/utils/test_split_data.py:24]: To be discussed in the upcoming dev meeting

File: src/equisolve/utils/split_data.py (hunk 4 above)
Comment (jwa7) [reply src/equisolve/utils/split_data.py:238]: Changed re: your suggestions; None doesn't shuffle, an int sets the seed for shuffling, -1 removed.

File: src/equisolve/utils/split_data.py (hunk 3 above)
Comment (jwa7) [reply src/equisolve/utils/split_data.py:244]: Function deleted and absorbed into the main fxn

File: src/equisolve/utils/split_data.py (hunk 1 above)
Comment (jwa7) [reply src/equisolve/utils/split_data.py:132]: Discussed in person.

**Comment by jwa7** [comment]
Changes made @PicoCentauri

**Review by PicoCentauri — APPROVED** [review]
Very nice!
